### ISSUE TITLE
HTML entity is erroneously converted

### ISSUE THREAD

**Issue by p3k** [commonmark/commonmark-java#307]
Steps to reproduce the problem (provide example Markdown if applicable):

```md
This endash: –  
should look like this endash: &#150;
```

Expected behavior:

```
This endash: –
should look like this endash: –
```

Actual behavior:

```
This endash: –
should look like this endash: 
```

The reference implementation at <https://spec.commonmark.org/dingus/> does behave differently, it displays the expected behavior.

Is this an abberration in the Java implementation?

**Comment by robinst** [comment]
Interesting case! So the spec only [says this](https://spec.commonmark.org/0.31.2/#decimal-numeric-character-references):

> Decimal numeric character references consist of `&#` + a string of 1--7 arabic digits + `;`.
> A numeric character reference is parsed as the corresponding Unicode character.
> Invalid Unicode code points will be replaced by the REPLACEMENT CHARACTER (`U+FFFD`).
> For security reasons, the code point `U+0000` will also be replaced by `U+FFFD`.

And that's what commonmark-java implements. `&#150;` is 150 in decimal, which is 96 in hex and so corresponds to the Unicode code point [U+0096](https://www.compart.com/en/unicode/U+0096), which is <Start of Guarded Area> (SPA).

But it turns out that commonmark.js implements this differently. It uses the `entities` library which has this replacement map: [decode_codepoint.ts](https://github.com/fb55/entities/blob/f312477409a36f1d3ffe0b8047a37aeb3c17303b/src/decode_codepoint.ts#L5).

As you can see there, 150 is mapped to 8211, which is [U+2013](https://www.compart.com/en/unicode/U+2013) aka En Dash. (If you use `&#8211;` in commonmark-java, it would work.)

___

Looking at the HTML spec (which commonmark usually tries to be compatible with), says [this](https://html.spec.whatwg.org/multipage/syntax.html#character-references):

> The numeric character reference forms described above are allowed to reference any code point excluding U+000D CR, [noncharacters](https://infra.spec.whatwg.org/#noncharacter), and [controls](https://infra.spec.whatwg.org/#control) other than [ASCII whitespace](https://infra.spec.whatwg.org/#ascii-whitespace).

(U+0096 *is* a control character)

And then in the [parsing section](https://html.spec.whatwg.org/multipage/parsing.html#numeric-character-reference-end-state):

> If the number is one of the numbers in the first column of the following table, then find the row with that number in the first column, and set the character reference code to the number in the second column of that row.

And there it is, the replacement table that `entities` uses (I assume it's the same, I haven't checked).

___

So, tl;dr:
* 150 in decimal maps to a control character in Unicode, and the HTML spec uses a different character for it (probably based on some legacy encoding).
* The spec doesn't mention the mapping
* commonmark.js implements the mapping
* commonmark-java doesn't

So technically it's not a bug, but I think it makes sense to follow HTML and commonmark.js here.

Note that GitHub which AFAIK uses cmark doesn't do the same replacement, it renders `&#150;` as: &#150;
(The code for that is probably [here](https://github.com/commonmark/cmark/blob/db0da216ba81435ce1547f9d1159a7f91298d187/src/houdini_html_u.c#L90).)

Next steps:

1. I'll raise an issue in https://github.com/commonmark/commonmark-spec
2. I'll work on adding the replacement table to commonmark-java

**Comment by p3k** [comment]
Wow, thanks for the deep insights @robinst 

> the HTML spec uses a different character for it (probably based on some legacy encoding)

I probably should have added that the corresponding content was created in the 1990s where encoding HTML entities like this was pretty common. The Markdown environment later grew around it.

So I see this could be kind of way-back backwards-compatibility 👴

And I noticed the issue on GitHub, too :relaxed:

**Comment by robinst** [comment]
> I probably should have added that the corresponding content was created in the 1990s

The Markdown content? 😱

I don't know if you've read the discussion on the spec issue, but I'm leaning towards closing this as "not a bug" for now until the spec says something different about numeric character references.

**Comment by robinst** [comment]
I see. Can you migrate the content to replace `&#150;` with `&#8211;` (or just a literal &#8211;)?

**Comment by p3k** [comment]
Sure, that is always a possibility, eventually.

But still, CommonMark’s behavior seems to be inconsistent in different implementations, at least when comparing the Java version to the one running the reference implementation.

**Comment by robinst** [comment]
> But still, CommonMark’s behavior seems to be inconsistent in different implementations, at least when comparing the Java version to the one running the reference implementation.

Yeah. But [cmark](https://github.com/commonmark/cmark) is also a reference implementation and behaves differently. commonmark-java is implementing the spec as it is currently written, so it's a spec issue which I [raised here](https://github.com/commonmark/commonmark-spec/issues/765). (To be super pedantic, commonmark.js is the one currently not following the spec.)

I'm going to close this one, we can reopen once there's a decision on the spec issue. Until then I recommend not using legacy numeric character references. Thanks for bringing this up again!
